File: py_code/142. Linked List Cycle II.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
      fast,slow =head,head
      if fast==None or fast.next==None: return None
      fast=fast.next.next
      slow=slow.next
      while fast!=slow :
        if fast==None or fast.next==None: return None
        fast=fast.next.next
        slow=slow.next
      fast=head
      while slow!=fast :
        slow=slow.next
        fast=fast.next
      return fast

      pass

File: js_code/142.Linked List Cycle II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
 function ListNode(val) {
     this.val = val;
     this.next = null;
}
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var detectCycle = function(head) {
     let fast=head,slow=head;
     if(fast==null || fast.next==null) return null;
     fast=fast.next.next;
     slow=slow.next;
     while(slow!=fast)
     {
       if(fast==null || fast.next==null) return null;
       fast=fast.next.next;
       slow=slow.next;
     }
     fast=head;
     while(slow!=fast)
     {
       fast=fast.next;
       slow=slow.next;
     }
     return fast;
};

// 好吧，js也有do....while语句可以使用